File: src/Managers/input_manager.py
"""
Input Manager - Orchestrates the input source (camera or video) and feeds frames into the buffer.

Runs a capture loop in a background thread:
    [Camera / Video] --read_frame()--> [FrameBuffer]
"""
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent))

import time
import threading
from typing import Optional, Union

from utils.config import Config
from utils.logger import Logger
from Handlers.camera_input_handler import CameraInputHandler
from Handlers.video_input_handler import VideoInputHandler
from Handlers.frame_buffer_handler import FrameBufferHandler


class InputManager:
    """
    Manages the input pipeline: source → buffer.
    
    Picks camera or video based on constructor args,
    runs a capture thread that reads frames and writes them into the shared buffer.
    """

    def __init__(self, config: Config, buffer: FrameBufferHandler, video_path: Optional[str] = None):
        """
        Args:
            config: Application configuration.
            buffer: Shared frame buffer that consumers read from.
            video_path: If provided, use video file instead of camera.
        """
        self.logger = Logger("InputManager")
        self.config = config
        self.buffer = buffer

        # Create the appropriate input source
        if video_path:
            self.source: Union[CameraInputHandler, VideoInputHandler] = VideoInputHandler(config, video_path)
            self._source_type = "video"
        else:
            self.source = CameraInputHandler(config)
            self._source_type = "camera"

        self._thread: Optional[threading.Thread] = None
        self._running = False

        # FPS control
        self._target_fps = self.config.get_int('camera.fps', 30)
        self._frame_interval = 1.0 / self._target_fps if self._target_fps > 0 else 0.033

    def start(self) -> bool:
        """Start the input source and begin the capture loop."""
        if self._running:
            self.logger.warning("Input manager already running")
            return True

        if not self.source.start():
            self.logger.error(f"Failed to start {self._source_type} source")
            return False

        self._running = True
        self._thread = threading.Thread(target=self._capture_loop, name="InputCapture", daemon=True)
        self._thread.start()

        self.logger.info(f"Input manager started ({self._source_type} @ {self._target_fps} fps)")
        return True

    def _capture_loop(self):
        """Background loop: read frames from source, push into buffer."""
        while self._running:
            loop_start = time.monotonic()

            frame = self.source.read_frame()

            if frame is not None:
                self.buffer.write_frame(frame)
            else:
                # Source returned None — either EOF (video) or transient error (camera)
                if not self.source.is_running:
                    self.logger.info("Input source ended")
                    self._running = False
                    break
                # Transient failure — brief pause then retry
                time.sleep(0.01)
                continue

            # FPS throttle: sleep the remainder of the frame interval
            elapsed = time.monotonic() - loop_start
            sleep_time = self._frame_interval - elapsed
            if sleep_time > 0:
                time.sleep(sleep_time)

    def stop(self):
        """Stop the capture loop and release the input source."""
        if not self._running:
            return
        self._running = False
        if self._thread:
            self._thread.join(timeout=2.0)
            self._thread = None
        self.source.stop()
        self.logger.info("Input manager stopped")

    @property
    def is_running(self) -> bool:
        return self._running

    @property
    def source_type(self) -> str:
        return self._source_type


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("--video", "-v", type=str, default=None, help="Path to video file")
    args = parser.parse_args()

    config = Config()
    buffer = FrameBufferHandler(config)
    mgr = InputManager(config, buffer, video_path=args.video)

    try:
        if mgr.start():
            # Let it run for 2 seconds then report
            time.sleep(2)
            print(f"✓ Source: {mgr.source_type}")
            print(f"  Buffer: {buffer.size}/{buffer.capacity} frames")
            print(f"  Duration: {buffer.duration_seconds:.2f}s")
        else:
            print("✗ Failed to start input")
    finally:
        mgr.stop()

File: src/managers/InputManager.cpp
#include "managers/InputManager.hpp"
#include "handlers/CameraHandler.hpp"
#include "handlers/VideoHandler.hpp"
#include "handlers/FrameBuffer.hpp"
#include "utils/EventBus.hpp"
#include "utils/Logger.hpp"
#include "utils/Config.hpp"
#include <chrono>

InputManager::InputManager(std::shared_ptr<FrameBuffer> buffer) 
    : m_Buffer(buffer), m_Running(false), m_FrameCounter(0) {
    
    // Abstract logic: if video path is provided in config, use VideoHandler, else CameraHandler
    std::string video_path = ""; // Config::getInstance().get<std::string>("video.test.path", "");
    
    if (!video_path.empty()) {
        m_Video = std::make_unique<VideoHandler>(video_path);
    } else {
        m_Camera = std::make_unique<CameraHandler>();
    }
}

InputManager::~InputManager() { stop(); }

void InputManager::start() {
    if (m_Running) return;
    
    bool started = false;
    if (m_Video) started = m_Video->start();
    else if (m_Camera) started = m_Camera->start();

    if (!started) {
        Logger::error("Input source failed to start.");
        return;
    }

    m_Running = true;
    m_Thread = std::thread(&InputManager::loop, this);
}

void InputManager::stop() {
    m_Running = false;
    if (m_Thread.joinable()) {
        m_Thread.join();
    }
    if (m_Video) m_Video->stop();
    if (m_Camera) m_Camera->stop();
}

void InputManager::loop() {
    Logger::info("Input Capture loop started.");
    
    while(m_Running) {
        cv::Mat frame;
        bool ret = false;
        
        if (m_Video) ret = m_Video->readFrame(frame);
        else if (m_Camera) ret = m_Camera->readFrame(frame);

        if (ret && !frame.empty()) {
            double ts = std::chrono::duration_cast<std::chrono::milliseconds>(
                std::chrono::system_clock::now().time_since_epoch()).count() / 1000.0;
                
            Core::SharedFrame sf{ std::make_shared<cv::Mat>(frame.clone()), ts, m_FrameCounter++ };
            
            if (m_Buffer) m_Buffer->write(sf);
            Core::EventBus::getInstance().publish(Core::EventType::FRAME_CAPTURED, sf);
        } else {
            // Sleep short duration if empty frame
            std::this_thread::sleep_for(std::chrono::milliseconds(5));
        }
    }
}
